### DISCUSSION TITLE
vscpl2drv-tcpipsrv : building from source error

### DISCUSSION THREAD

**Discussion by dinguluer** [grodansparadis/vscp#275] [Q&A] [answered]
Hi,

I am following link to build from source on my Linux desktop :--
https://github.com/grodansparadis/vscpl2drv-tcpipsrv

git clone --recurse-submodules -j8 https://github.com/grodansparadis/vscp.git
git clone --recurse-submodules -j8 https://github.com/grodansparadis/vscpl2drv-tcpipsrv.git 
cd vscpl2drv-tcpipsrv
mkdir build
cd build
cmake ..
make

![image](https://github.com/grodansparadis/vscp/assets/370819/90c4790f-2bc2-4b48-b511-3bc1daddd3aa)

Regards,
Dinesh

**Comment by dinguluer** [comment, marked as answer]
Please anyone who can suggest on this query ?

Regards,
Dinesh

**Reply by grodansparadis** [reply]
You have to options here. 

1.) Set dependency's to the release version for vscp. That is checkout 15.,0.0 from the repo instead of the latest code and link against the release version instead.
2.)   VSCP_REMOTE_VARIABLE_CODE_JSON is removed in later versions as remote variables in the daemon now use the same types as remote variables for embedded news t(previously called abstractions) o avoid confusion.  So you can also just comment away the line this is on or change it to VSCP_REMOTE_VARIABLE_CODE_STRING. The code is not currently used anyway.

Why do you not just use the install package?

ps Also a matter of order. It is OK to use the discussions to ask questions like this. But the right way is to open a ticket in the vscpl2drv-tcpipsrv repository. That way you help others asking the same question as you are. Still asking here is better then a direct email to me of course (which is really BAD practice. ds

**Reply by grodansparadis** [reply]
BTW

If uou are just looking for a tcp/ip interface Maartens tool uvscpd can be useful  https://github.com/mzanders/uvscpd

**Reply by dinguluer** [reply]
>> uvscpd


What will it do ?

**Reply by dinguluer** [reply]
>>1.) Set dependency's to the release version for vscp. That is checkout 15.,0.0 from the repo instead of the latest code and link against the release version instead.

Which command to use to checkout 15.,0.0 from the repo  ?

**Reply by grodansparadis** [reply]
uvscp is a simpler server that connect to CAN and export a VSCP tcp/ip interface

[git checkout ](https://www.xmodulo.com/how-to-checkout-specific-version-of-git-repository.html)

**Reply by dinguluer** [reply]
You told :--
>> 1.) Set dependency's to the release version for **vscp**. That is **checkout 15.,0.0 from the repo** instead of the latest code and link against the release version instead.

*Use git checkout, command to checkout 15.,0.0*  which repository version 15.0.0 to be checked, out of these two, VSCP ? :--
https://github.com/grodansparadis/vscp.git
https://github.com/grodansparadis/vscpl2drv-tcpipsrv.git

**Reply by dinguluer** [reply]
I took the log of the vscp repository :--

![image](https://github.com/grodansparadis/vscp/assets/370819/d9ba11d5-c6cd-472f-92d9-652ea19b664c)


Please can you confirm this commit is right one to take :--

![image](https://github.com/grodansparadis/vscp/assets/370819/faa9fc51-c544-432d-b597-73ac0c00cbbf)

**Comment by dinguluer** [comment]
Hi Ake,

Please suggest on it.

Regards,
Dinesh

**Comment by grodansparadis** [comment]
You should match vscp with the driver.

**Comment by grodansparadis** [comment]
![image](https://github.com/grodansparadis/vscp/assets/1358883/1df5a462-a3f6-4715-b196-2767f5796601)

**Reply by dinguluer** [reply]
If I am correct this is the commit number, **right ?** :--
[commit number ](https://github.com/grodansparadis/vscp/commit/a3f29b38c6b7c8d25265091b295727f55c327d05)

And i shall use this command :--
**git commit a3f29b38c6b7c8d25265091b295727f55c327d05**

![image](https://github.com/grodansparadis/vscp/assets/370819/4783f89f-5d34-4e65-b143-9762ea47f554)

**Comment by grodansparadis** [comment]
git checkout ....

**Reply by dinguluer** [reply]
Ok compiled successfully.

Thanks.


![image](https://github.com/grodansparadis/vscp/assets/370819/6e0f068c-1a27-4f9e-ba79-a511fa6587e7)

**Reply by dinguluer** [reply]
sudo make install

is working

![image](https://github.com/grodansparadis/vscp/assets/370819/a8e66716-dc42-4e2d-872d-c8c1fc7e55b1)

**Comment by dinguluer** [comment]
**telnet localhost 9598  .. not working ..**

Please suggest how to resolve it ?

I use kill command to end the vscpd :--
![image](https://github.com/grodansparadis/vscp/assets/370819/f39872dc-7d17-496d-8e26-abff1bb84c42)

I even used this command to make sure vscpd is restarted :--
sudo systemctl restart vscpd


Then vscpd restarted as it is system service :---
![image](https://github.com/grodansparadis/vscp/assets/370819/cd6f0e10-b810-40df-9bef-cd0c1b82be6e)


Still TELNET not working :--
![image](https://github.com/grodansparadis/vscp/assets/370819/30cfa84a-9803-4b2f-b685-d44dd85108dc)

vscpl2drv-tcpipsrv driver JSON files are here :--
![image](https://github.com/grodansparadis/vscp/assets/370819/6b122544-57c4-495f-9161-d91788b67bdd)

My drivers are here :---
![image](https://github.com/grodansparadis/vscp/assets/370819/a4398f14-a1bb-4f74-8f16-8dcde826100b)


My Driver JSON files are here :--
[DRIVER_JSON_FILE.zip](https://github.com/grodansparadis/vscp/files/12814278/DRIVER_JSON_FILE.zip)


This is my vscpd.json file :--
[vscpd.zip](https://github.com/grodansparadis/vscp/files/12814222/vscpd.zip)

**Comment by grodansparadis** [comment]
Is the driver started accordning to the logs?

**Reply by dinguluer** [reply]
Which LOG, this one ?

![image](https://github.com/grodansparadis/vscp/assets/370819/b3f6310f-607f-4297-b556-385f809ac6a2)

**Reply by dinguluer** [reply]
Please suggest which log to look ?

Regards,
Dinesh

**Reply by dinguluer** [reply]
Hi Ake,

Please can you reply to my query ?

Regards,
DInesh

**Comment by grodansparadis** [comment]
Sorry for late reply, have been on the roads.

You have two logs. The vscpd log and the log for trhe driver itself.  First check the vscpd log so the driver started. If it did check for problems in the tcpipsrv drivers log.

**Reply by dinguluer** [reply]
Thanks for your reply.

>> First check the vscpd log so the driver started. 

where to check this ? Which path in Linux machine ?

>> If it did check for problems in the tcpipsrv drivers log.

where to check this ? Which path in Linux machine ?

**Comment by grodansparadis** [comment]
You refered to the vscpd log file above. They are by default at var/log/vscp/..

**Reply by dinguluer** [reply]
I deleted all the log files in :--
/var/log/vscp

I restarted the vscpd :--
![image](https://github.com/grodansparadis/vscp/assets/370819/a03c20eb-9c78-4d1b-a3e7-3c322e94d06f)

Then check the status :--
sudo systemctl status vscpd
![image](https://github.com/grodansparadis/vscp/assets/370819/041a6260-2815-4f81-942d-cdc7a5b4f430)

 log files ls recreated here in :--
/var/log/vscp
![image](https://github.com/grodansparadis/vscp/assets/370819/283cd80d-454e-41d8-a760-1e096852e319)


**But log file is empty, nothing is inside, vscpd.log**




Telnet connection still not working :--
![image](https://github.com/grodansparadis/vscp/assets/370819/7a911122-ac0d-4c64-b0ae-81849b9895ff)


Please suggest how to resolve it ?

Regards,
DInesh

**Comment by grodansparadis** [comment]
Do you get any messages on the MQTT server from the daemon (heartbeats)?
Can you post the config file.

**Reply by dinguluer** [reply]
Thanks for your reply.

>> Do you get any messages on the MQTT server from the daemon (heartbeats)?
>> Can you post the config file.

How to check this ? 

Please suggest.

Regards,
Dinesh

**Comment by grodansparadis** [comment]
Config file: /etc/vscp/vscpd.json
Heartbeats:  Configured MQTT server on configured topic. Default  vscp-daemon/{{guid}} where {{guid}}is configured guid and vscp-/{{guid}}

**Reply by dinguluer** [reply]
>>Heartbeats: Configured MQTT server on configured topic. Default vscp-daemon/{{guid}} where {{guid}}is configured guid and vscp-/{{guid}}

I did not got what you want to say here ?
What exactly I have to do at my end to make this driver working ?
Please suggest.

I have attached my file vscpd.json.
[vscpd.json](https://github.com/grodansparadis/vscp/files/13178253/vscpd.json)

This is setting for my vscpl2drv-tcpipsrv driver :---

`      {
        "enable" : true,
        "name" : "tcpip-srv",
        "path-driver" : "/var/lib/vscp/drivers/level2/libvscpl2drv-tcpipsrv.so",
        "path-config" : "/usr/share/vscpl2drv-tcpipsrv/tcpipsrv.json",
        "guid" : "FF:FF:FF:FF:FF:FF:FF:F5:02:88:88:00:00:00:00:01",
      
        "mqtt": {
          "bind": "",
          "host": "test.mqtt.org",
          "port": 1883,
          "mqtt-options": {
            "tcp-nodelay": true,
            "protocol-version": 311,
            "receive-maximum": 20,
            "send-maximum": 20,
            "ssl-ctx-with-defaults": 0,
            "tls-ocsp-required": 0,
            "tls-use-os-certs": 0
          },
          "user": "vscp",
          "password": "secret",
          "clientid": "the-vscp-daemon tcpipsrv driver",
          "publish-format": "json",
          "subscribe-format": "auto",
          "qos": 1,
          "bcleansession": false,
          "bretain": false,
          "keepalive": 60,
          "bjsonmeasurementblock": true,
          "reconnect": {
            "delay": 2,
            "delay-max": 10,
            "exponential-backoff": false
          },
          "tls": {
            "cafile": "",
            "capath": "",
            "certfile": "",
            "keyfile": "",
            "pwkeyfile": "",
            "no-hostname-checking": true,
            "cert-reqs": 0,
            "version": "",
            "ciphers": "",
            "psk": "",
            "psk-identity": ""
          },
          "will": {
            "topic": "vscp-daemon/{{srvguid}}/will",
            "qos": 1,
            "retain": true,
            "payload": "VSCP Daemon is down"
          },
          "subscribe" : [
            {
              "topic": "vscp/tcpipsrv/{{guid}}/#",
              "qos": 0,
              "v5-options": 0,
              "format": "auto"
            }
          ],
          "publish" : [
            {
              "topic": "vscp/{{guid}}/{{class}}/{{type}}/{{nodeid}}",
              "qos": 1,
              "retain": false,
              "format": "json"
            }
          ]
        }
      },`

**Comment by grodansparadis** [comment]
Looks OK as far as I can see
If you stop the service and run vscpd from the user prompt "vscpd -s" what output do you get?

**Reply by dinguluer** [reply]
>> If you stop the service 

I used this command to stop the service :--
sudo systemctl stop vscpd

Once service is stopped i tried to check if it is stopped or not :--
$ sudo systemctl status vscpd
![image](https://github.com/grodansparadis/vscp/assets/370819/792d5e07-df33-4519-9c67-3629fb99c737)



>> and run vscpd from the user prompt "vscpd -s" what output do you get?

![image](https://github.com/grodansparadis/vscp/assets/370819/a3068845-8ae7-4014-a0c9-7700edbe8fc1)

Then I again check the status of vscpd :--
![image](https://github.com/grodansparadis/vscp/assets/370819/6940c270-1bb9-4a63-937a-220ef5c027ec)


Please suggest.

Regards,
Dinesh

**Comment by grodansparadis** [comment]
You must run vscp as the vscp user or as root. Thats why it does not start for you. Preferebly root if you start it in stand alone mode.

**Reply by dinguluer** [reply]
I enter root login :--
su -i

Then run following command this is status :--
vscpd -s
![image](https://github.com/grodansparadis/vscp/assets/370819/aad51961-2ba4-4b09-82fa-598258a1bdbe)


Then after sometime this is status :--
![image](https://github.com/grodansparadis/vscp/assets/370819/cbea21c2-563a-4c45-9ae9-7c6f6d0f0dec)

**Comment by grodansparadis** [comment]
That says it can't connect to the MQTT server. Check server address, uisername, password for the MQTT server you use

**Reply by dinguluer** [reply]
I use this MQTT server as recommended by VSCPD  :--
libmosquitto

>> Check server address, uisername, password for the MQTT server you use

I using libmosquitto first time . How to configure it in libmosquitto ?

Please suggest.

Regards,
Dinesh

**Comment by grodansparadis** [comment]
libmosquitto is the library for programmers. mosquitto is the server.

There is a lot available about Mosquitto if you check around on the net. For instance `https://mqtt.org/getting-started/

You can also use test.mosquitto.org as your server.

**Comment by grodansparadis** [comment]
btw if you want to look at real live VSCP data you can subscribe to "vscp/#"  on [test.mosquitto.org](url)  It is live data from my house/office.   Described [](here https://github.com/grodansparadis/vscp/wiki/Demo)   Electrical usage is for instance displayed live from this data [here](https://demo.vscp.org/power.html)

**Comment by dinguluer** [comment]
Thanks for your reply.

**libmosquitto-dev, is used for client application development, right ? :--**
https://packages.ubuntu.com/focal/libmosquitto-dev
MQTT version 5.0/3.1.1/3.1 client library, development files

Earlier i thought installing libmosquitto-dev, installed the Mosquitto server :---
https://grodansparadis.github.io/vscp/#/setting_up_the_system?id=building-binaries-from-source
sudo apt install git libssl-dev libsystemd-dev libexpat-dev libwrap0-dev libz-dev libmosquitto-dev

**This link explains what is MQTT topic, if i am correct "vscp/#" follow same rule ? :---**
What Are MQTT Topics and Their Role in MQTT Message Filtering?
https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/

**Do we have any link where details of "vscp/#" VSCP topics are mentioned ?**

**Use this WEB server from mosquitto on Internet for MQTT broker, right ? :--**
https://test.mosquitto.org/

**To install local host on linux, follow this process, right :--**
http://www.steves-internet-guide.com/install-mosquitto-linux/

> btw if you want to look at real live VSCP data you can subscribe to "vscp/#" on test.mosquitto.org

How you updated this TOPIC on "vscp/#" on test.mosquitto.org ?
**Do we need to configure MQTT broker at test.mosquitto.org for publish/subscribe topics "vscp/#" ?**

**Please share link** where it is mentioned **how we can configure, MQTT broker at test.mosquitto.org** for publish/subscribe topics like **"DINESH/#" .**

Regards,
Dinesh

**Reply by dinguluer** [reply]
**Here is source code for mosquitto server right ? :--**
https://github.com/eclipse/mosquitto
https://github.com/eclipse/mosquitto#installing

**If i need to build mosquitto server for my Specific SOC then need to follow this process, right ?**
https://github.com/eclipse/mosquitto#building-from-source


**Installing on local Linux PC mosquitto server, this is right process right ?**
https://github.com/eclipse/mosquitto#installing
https://mosquitto.org/download/
**snap install mosquitto**

**Comment by grodansparadis** [comment]
Yes correct,   The Mosquitto server is usually also available as a binary packet on most systems (Windows, Ubuntu, Debian etc.) Easier to install it that way.  But no problem to build on your own either of course. In the past one have to do that to get websocket support but now it is included in the standard packages.

**Reply by dinguluer** [reply]
**How you updated this TOPIC** on "vscp/#" on test.mosquitto.org ?
Do we need to configure MQTT broker at test.mosquitto.org for publish/subscribe topics "vscp/#" ?

How we can configure, MQTT broker at test.mosquitto.org for publish/subscribe topics like **"DINESH/#" .**

Please suggest.

Regards,
Dinesh

**Comment by grodansparadis** [comment]
It's publisher/subscriber based so subscribing to the topic will get evterhing that is published to it by other clients. The "#" is a wildcard meaning all levels below  so you get everything.  Test on tes@mosquitto.org on this suject with MQTT explorer or simular.

For a "DINESH" topic you just publish to it. No "#" wildcard for publish.

**Reply by dinguluer** [reply]
Thanks for the reply

Ok got so it means any thing we can publish to test.mosquitto.org broker, it will accept it.
I was thinking we might have to take some permission from test.mosquitto.org Broker to get right to publish.

Where in VSCPD config files we have to specify which broker to use, this file, **right** ? :--
/etc/vscp/vscpd.json

Here i have  specified the test.mqtt.org :--
![image](https://github.com/grodansparadis/vscp/assets/370819/1c5ecc0d-a9a9-4e78-b1c2-2ea2431ba784)

**Shall i change it to this ? :--**
test.mosquitto.org

Please confirm.

Regards,
Dinesh

**Comment by grodansparadis** [comment]
Yes you can use test.mosquitto.org but only for tests as it can be down anytime.  If you need a more reliable MQTT server install one of your own.   

Aloso join us here https://www.facebook.com/groups/980614973001987

**Reply by dinguluer** [reply]
>> If you need a more reliable MQTT server install one of your own.

Install on Local Host or OWN web-server ?

Regards,
Dinesh

**Reply by grodansparadis** [reply]
Loclahost i simplest. Own server convenient. Happy Diwali.

**Comment by dinguluer** [comment]
Thanks and Happy Diwali to you too !!!
